File: 1930_uniqueLength3PalindromicSubsequences/main.py
from typing import List
import collections
class Solution:
    def countPalindromicSubsequence(self, s: str) -> int:
        res = set()
        left = set()
        right = collections.Counter(s)

        for i in range(len(s)):
            right[s[i]] -= 1
            if right[s[i]] == 0:
                right.pop(s[i])
            
            for j in range(26):
                c = chr(ord('a') + j)
                if c in left and c in right:
                    res.add((s[i], c))
            
            left.add(s[i])
        
        return len(res)
    
class Solution(object):
    def countPalindromicSubsequence(self, s):
        d=collections.defaultdict(list)
        for i,c in enumerate(s):
            d[c].append(i)
        ans=0
        for el in d:
            if len(d[el])<2:
                continue
            a=d[el][0]
            b=d[el][-1]
            ans+=len(set(s[a+1:b]))
        return(ans)

File: 1930_uniqueLength3PalindromicSubsequences/main.cpp
#include <string>
#include <unordered_map>
#include <unordered_set>

class Solution {
public:
    int countPalindromicSubsequence(std::string s) {
        
        std::unordered_map<int,int> right(26);
        std::unordered_set<int> left;
        std::unordered_set<std::string> res;
        for(int i=0;i<s.size();i++)
          right[s[i]-'a']++;
       
        for(int i = 0; i < s.size(); ++i){
            --right[s[i]-'a'];
            if(right[s[i]-'a'] == 0)
                right.erase(s[i]-'a');
            
            if(left.size() > 0){
                for(auto it:left){
                    std::string temp="";
                    if(right.find(it) != right.end()){
                        temp += it;
                        temp += s[i];
                        temp += it;
                        res.insert(temp);
                    }   
                }   
            }
            left.insert(s[i]-'a');
        }
        
        return res.size();
    }
};